File: shared/src/types/meeting.ts
export type MeetingRoom = {
  id: string;
  name: string;
  location: string | null;
  capacity: number | null;
  description: string | null;
  is_active: boolean;
  created_at: string | null;
  updated_at: string | null;
};

export type ReservationStatus = "CONFIRMED" | "CANCELLED";

export type ReservationParticipant = {
  developer_id: string;
  name: string | null;
  email: string | null;
};

export type MeetingReservation = {
  id: string;
  room_id: string;
  room_name: string | null;
  organizer_id: string | null;
  organizer_name: string | null;
  title: string;
  description: string | null;
  start_at: string;
  end_at: string;
  status: ReservationStatus;
  participants: ReservationParticipant[];
  created_at: string | null;
  updated_at: string | null;
};

export type DirectoryUser = {
  id: string;
  email: string;
  name: string;
  role: string;
  is_active: boolean;
};

// 임직원 디렉토리 엔트리 — 회의 참석자 선택 등 people-picker 용.
// email 은 Slack 알림 대상 (company_email ?? personal_email).
export type DirectoryMember = {
  id: string;
  name: string;
  email: string | null;
  employment_type: string;
};

File: backend/app/schemas/meeting.py
"""Pydantic schemas for 회의실 예약(meetings)."""

from datetime import datetime
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, model_validator

ReservationStatus = Literal["CONFIRMED", "CANCELLED"]


# ---------------------------------------------------------------------------
# Room
# ---------------------------------------------------------------------------


class MeetingRoomCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    location: Optional[str] = Field(default=None, max_length=200)
    capacity: Optional[int] = Field(default=None, ge=0)
    description: Optional[str] = None
    is_active: bool = True


class MeetingRoomUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    location: Optional[str] = Field(default=None, max_length=200)
    capacity: Optional[int] = Field(default=None, ge=0)
    description: Optional[str] = None
    is_active: Optional[bool] = None


class MeetingRoomOut(BaseModel):
    id: UUID
    name: str
    location: Optional[str] = None
    capacity: Optional[int] = None
    description: Optional[str] = None
    is_active: bool
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Reservation
# ---------------------------------------------------------------------------


class ReservationParticipantOut(BaseModel):
    developer_id: UUID
    name: Optional[str] = None
    email: Optional[str] = None


class MeetingReservationCreate(BaseModel):
    room_id: UUID
    title: str = Field(min_length=1, max_length=200)
    description: Optional[str] = None
    start_at: datetime
    end_at: datetime
    participant_developer_ids: list[UUID] = Field(default_factory=list)

    @model_validator(mode="after")
    def _check_range(self) -> "MeetingReservationCreate":
        if self.end_at <= self.start_at:
            raise ValueError("end_at 은 start_at 보다 커야 합니다.")
        return self


class MeetingReservationUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = None
    start_at: Optional[datetime] = None
    end_at: Optional[datetime] = None
    participant_developer_ids: Optional[list[UUID]] = None
    status: Optional[ReservationStatus] = None

    @model_validator(mode="after")
    def _check_range(self) -> "MeetingReservationUpdate":
        if self.start_at and self.end_at and self.end_at <= self.start_at:
            raise ValueError("end_at 은 start_at 보다 커야 합니다.")
        return self


class MeetingReservationOut(BaseModel):
    id: UUID
    room_id: UUID
    room_name: Optional[str] = None
    organizer_id: Optional[UUID] = None
    organizer_name: Optional[str] = None
    title: str
    description: Optional[str] = None
    start_at: datetime
    end_at: datetime
    status: ReservationStatus
    participants: list[ReservationParticipantOut] = Field(default_factory=list)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True
